executor/admin.go

```diff
@@ -457,3 +459,248 @@ func (e *RecoverIndexExec) NextChunk(ctx context.Context, chk *chunk.Chunk) erro
+				return nil
+			}
+		}
+	}
+}
+
+// NextChunk implements the Executor NextChunk interface.
+func (e *CleanupIndexExec) NextChunk(ctx context.Context, chk *chunk.Chunk) error {
+	chk.Reset()
+	if e.done {
+		return nil
+	}
+	for {
+		errInTxn := kv.RunInNewTxn(e.ctx.GetStore(), true, func(txn kv.Transaction) error {
+			err := e.fetchIndex(ctx, txn)
+			if err != nil {
+				return errors.Trace(err)
+			}
+			values, err := e.batchGetRecord(txn)
+			if err != nil {
+				return errors.Trace(err)
+			}
+			err = e.deleteDanglingIdx(txn, values)
+			if err != nil {
+				return errors.Trace(err)
+			}
+			return nil
+		})
+		if errInTxn != nil {
+			return errors.Trace(errInTxn)
+		}
+		if e.scanRowCnt == 0 {
+			break
+		}
+		e.scanRowCnt = 0
+		e.batchKeys = e.batchKeys[:0]
+		for k := range e.matchedIndex {
+			delete(e.matchedIndex, k)
+		}
+	}
+	e.done = true
+	chk.AppendUint64(0, e.removeCnt)
+	return nil
+}
+
+func (e *CleanupIndexExec) buildIndexScan(ctx context.Context, txn kv.Transaction) (distsql.SelectResult, error) {
+	dagPB, err := e.buildIdxDAGPB(txn)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	sc := e.ctx.GetSessionVars().StmtCtx
+	var builder distsql.RequestBuilder
+	ranges := ranger.FullNewRange()
+	kvReq, err := builder.SetIndexRanges(sc, e.table.Meta().ID, e.index.Meta().ID, ranges).
+		SetDAGRequest(dagPB).
+		SetKeepOrder(true).
+		SetFromSessionVars(e.ctx.GetSessionVars()).
+		Build()
+	kvReq.KeyRanges[0].StartKey = kv.Key(e.lastIdxKey).PrefixNext()
+	kvReq.Concurrency = 1
+	result, err := distsql.Select(ctx, e.ctx, kvReq, e.getIdxColTypes(), statistics.NewQueryFeedback(0, nil, 0, false))
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	result.Fetch(ctx)
+	return result, nil
+}
+
+// Open implements the Executor Open interface.
+func (e *CleanupIndexExec) Open(ctx context.Context) error {
+	if err := e.baseExecutor.Open(ctx); err != nil {
+		return errors.Trace(err)
+	}
+	e.idxChunk = chunk.NewChunk(e.getIdxColTypes())
+	e.matchedIndex = make(map[int64]idxData, e.batchSize)
+	e.batchKeys = make([]kv.Key, 0, e.batchSize)
+	e.idxValsBufs = make([][]types.Datum, e.batchSize)
+	sc := e.ctx.GetSessionVars().StmtCtx
+	idxKey, _, err := e.index.GenIndexKey(sc, []types.Datum{{}}, math.MinInt64, nil)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	e.lastIdxKey = idxKey
+	return nil
+}
+
+func (e *CleanupIndexExec) buildIdxDAGPB(txn kv.Transaction) (*tipb.DAGRequest, error) {
+	dagReq := &tipb.DAGRequest{}
+	dagReq.StartTs = txn.StartTS()
+	dagReq.TimeZoneOffset = timeZoneOffset(e.ctx)
+	sc := e.ctx.GetSessionVars().StmtCtx
+	dagReq.Flags = statementContextToFlags(sc)
+	for i := range e.idxCols {
+		dagReq.OutputOffsets = append(dagReq.OutputOffsets, uint32(i))
+	}
+
+	execPB := e.constructIndexScanPB()
+	dagReq.Executors = append(dagReq.Executors, execPB)
+	err := plan.SetPBColumnsDefaultValue(e.ctx, dagReq.Executors[0].IdxScan.Columns, e.idxCols)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
```

Comment:
Can these codes be extracted into a function？
The logic is used in many places.

Reply:
They looked similar, but not exactly the same. It needs a refactor. Maybe in the next pr.

Reply:
It's the same as `CheckIndexRangeExec`'s buildDAGPB.
But if you want to refactor some of the exec in admin, you can handle it in the next PR.